### PR TITLE
Native automatic mixed precision training support

### PR DESCRIPTION
**Pull request by BloodAxe**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (https://github.com/catalyst-team/catalyst/issues/740)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Description

This PR brings support of `torch.cuda.amp` (https://pytorch.org/docs/stable/notes/amp_examples.html) to catalyst, which is faster replacement of NVidia Apex. Instead of `apex: True` we use `amp: True` and that's all.

## Related Issue

https://github.com/catalyst-team/catalyst/issues/740

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/utils/distributed.py [hunk 1]
```diff
@@ -48,6 +48,15 @@ def check_apex_available() -> bool:
         return False and env_apex
 
 
+def check_amp_available() -> bool:
+    """Checks if torch.amp is available."""
+    major, minor = torch.__version__.split(".")[:2]
+    major = int(major)
+    minor = int(minor)
+    # torch.amp available since version 1.6 and above
+    return (major >= 1 and minor >= 6) or major > 1
```
Comment (Scitator) [inline catalyst/utils/distributed.py]: could you please use something like https://github.com/catalyst-team/catalyst/blob/9f97ad91c82322a18ac6a5764aaded7fc54720a4/catalyst/utils/seed.py#L27 ?

File: catalyst/core/callbacks/optimizer.py [hunk 2]
```diff
@@ -222,4 +224,166 @@ def on_batch_end(self, runner: IRunner) -> None:
             self._accumulation_counter = 0
 
 
-__all__ = ["OptimizerCallback"]
+class AMPOptimizerCallback(Callback):
```
Comment (Scitator) [inline catalyst/core/callbacks/optimizer.py:227]: what do you think about adding something like
`IOptimizercallback` - interface for optimizer callbacks and `OptimizerCallback` & `AMPOptimizerCallback` as instances

or maybe add some parameter to `OptimizerCallback` to choose correct `optimizer_step`?

**Review by Scitator — COMMENTED** [review]
btw, do you have any difficulties with codestyle?

File: catalyst/utils/distributed.py (hunk 1 above)
Comment (BloodAxe) [reply catalyst/utils/distributed.py]: Thanks, this way more cleaner way of checking version
